Generate reference doc for 1.21.0-rc.0 tag

reylejano opened the pull request:
Update generated reference API docs with tag 1.21.0-rc.0

/assign @kbhawkey 
/hold for review

irvifa reviewed (commented):
/lgtm

kbhawkey commented:
Hi @reylejano .
Were there any errors? I don't see any generated component/tool pages?
You should update the base reference page (to link to API reference v1.21)?
This link is referring to the v1.20 reference page.
I assume that the `kubectl` tool did not change (commands).

Preview page:
https://deploy-preview-27248--kubernetes-io-vnext-staging.netlify.app/docs/reference/
https://deploy-preview-27248--kubernetes-io-vnext-staging.netlify.app/docs/reference/generated/kubernetes-api/v1.21/
https://deploy-preview-27248--kubernetes-io-vnext-staging.netlify.app/docs/reference/generated/kubernetes-api/v1.20/
https://deploy-preview-27248--kubernetes-io-vnext-staging.netlify.app/docs/reference/generated/kubectl/kubectl-commands

Hold for review
/hold

kbhawkey commented:
Related PR (1.21 API reference pages):  https://github.com/kubernetes/website/pull/26413

kbhawkey commented:
Hi. I tried to build the component tool reference pages. The build failed (running script and running from command line). 
@tengqm I think you recently changed the build of gen-compdocs. Would you have time to sort out the build/module errors?
Looks like Kubernetes requires Go 1.16+ now. Thanks. I'll continue to look into the errors.

reylejano commented:
> Hi @reylejano .
> Were there any errors? I don't see any generated component/tool pages?
> You should update the base reference page (to link to API reference v1.21)?
> This link is referring to the v1.20 reference page.
> I assume that the `kubectl` tool did not change (commands, not descriptions).
> 
> Preview page:
> https://deploy-preview-27248--kubernetes-io-vnext-staging.netlify.app/docs/reference/
> https://deploy-preview-27248--kubernetes-io-vnext-staging.netlify.app/docs/reference/generated/kubernetes-api/v1.21/
> https://deploy-preview-27248--kubernetes-io-vnext-staging.netlify.app/docs/reference/generated/kubernetes-api/v1.20/
> https://deploy-preview-27248--kubernetes-io-vnext-staging.netlify.app/docs/reference/generated/kubectl/kubectl-commands
> 
> Hold for review
> /hold

@kbhawkey I did not receive any errors, I'll update the base reference page to link to API reference v1.21

reylejano commented:
@kbhawkey I updated my golang and re-ran. The build failed on the second run

sftim reviewed (commented):
/approve

kbhawkey commented:
👋 @reylejano .
It is not clear to me why the tool pages cannot build with the release candidates.
There appears to be some differences in the packaging/modules of the libraries (dependencies of k8s.io)?
Hoping @tengqm will chime in here.

tengqm commented:
@reylejano @kbhawkey I don't think we need/want to build the component reference at this stage. As you may know, to ensure a stable/repeatable generation of the references, we have switched to manage Go dependencies using go mod. The side effect is that we will not be able to generate the proper reference until the upstream repos are ready. By "ready", I mean the proper tags show up in those repos. For example, we will need the `v0.21.0` tag from the `k8s.io/api` repo, and that will not happen before 1.21.0 is released from `k8s.io/kubernetes`.

irvifa commented:
In this case then we can wait until the tag is published and then preparing for the document to be generated?

reylejano commented:
@kbhawkey @tengqm @irvifa , we discussed this PR a little bit on today's SIG Docs call. 
Do you think we should change the [release docs handbook timeline](https://github.com/kubernetes/sig-release/blob/master/release-team/role-handbooks/docs/Release-Timeline.md#generate-the-reference-documentation) to have the reference doc generation to the release day when the `v0.21.0` tag is available

kbhawkey commented:
Hi. In the past releases, we have created the reference documentation in time for the release and used the
last release candidate tag (sometimes even beta for testing). This worked well and enabled the docs site to include the generated reference documentation on the `dev-<release>` branch.
It is surprising to me that this has changed.
Creating the docs from the release candidate tags, enables us to evaluate and update the reference docs (or steps to generate the docs) in time for the release. I would think that the docs could use the release candidate tags (?).

Why is the `v1.21.0` (v0.21.0?) tag required?
` For example, we will need the v0.21.0 tag from the k8s.io/api repo, and that will not happen before 1.21.0 is released from k8s.io/kubernetes`

tengqm commented:
> Why is the v1.21.0 (v0.21.0?) tag required?

See https://github.com/kubernetes-sigs/reference-docs/blob/master/gen-compdocs/go.mod#L22-L46

kbhawkey commented on go.mod:
```diff
@@ -1,10 +1,10 @@
 require (
 	k8s.io/apimachinery v0.20.0
-	k8s.io/kubernetes v1.20.0
+	k8s.io/kubernetes v1.21.0
```
@reylejano , Are the go.mod changes related to the generated files in this PR?

reylejano replied:
@kbhawkey , the generated files are with the go.mode changes with line 7 in go.mod `k8s.io/kubernetes v1.21.0`
But I did not change go.mod lines 11-37:

> replace (
	k8s.io/api => k8s.io/api v0.20.0
	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.20.0
	k8s.io/apimachinery => k8s.io/apimachinery v0.20.0
	k8s.io/apiserver => k8s.io/apiserver v0.20.0
	k8s.io/cli-runtime => k8s.io/cli-runtime v0.20.0
	k8s.io/client-go => k8s.io/client-go v0.20.0
	k8s.io/cloud-provider => k8s.io/cloud-provider v0.20.0
	k8s.io/cluster-bootstrap => k8s.io/cluster-bootstrap v0.20.0
	k8s.io/code-generator => k8s.io/code-generator v0.20.0
	k8s.io/component-base => k8s.io/component-base v0.20.0
	k8s.io/component-helpers => k8s.io/component-helpers v0.20.0
	k8s.io/controller-manager => k8s.io/controller-manager v0.20.0
	k8s.io/cri-api => k8s.io/cri-api v0.20.0
	k8s.io/csi-translation-lib => k8s.io/csi-translation-lib v0.20.0
	k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.20.0
	k8s.io/kube-controller-manager => k8s.io/kube-controller-manager v0.20.0
	k8s.io/kube-proxy => k8s.io/kube-proxy v0.20.0
	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.20.0
	k8s.io/kubectl => k8s.io/kubectl v0.20.0
	k8s.io/kubelet => k8s.io/kubelet v0.20.0
	k8s.io/kubernetes => k8s.io/kubernetes v1.20.0
	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.20.0
	k8s.io/metrics => k8s.io/metrics v0.20.0
	k8s.io/mount-utils => k8s.io/mount-utils v0.20.0
	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.20.0
	k8s.io/sample-cli-plugin => k8s.io/sample-cli-plugin v0.20.0
	k8s.io/sample-controller => k8s.io/sample-controller v0.20.0
)

I can re-run with changes in lines 11-37 in go.mod from `v0.20.0` to `v0.21.0` and add to this PR

tengqm replied:
The problem is that these modules are staged from the k/k repo. When `1.21.0` is not out, you won't get the `v0.21.0` tag from these modules.

reylejano replied:
> The problem is that these modules are staged from the k/k repo. When `1.21.0` is not out, you won't get the `v0.21.0` tag from these modules.

@tengqm that's what I thought, but I re-ran with changes to go.mod to `v0.21.0 `or `v1.21.0`, I'll push but won't squash yet to see the changes

kbhawkey commented:
/hold cancel

reylejano commented:
@kbhawkey @tengqm Pushed commit a2eed93 after re-running with complete version/tag changes in go.mod. There were significant changes. 

Let me know if this is looks better or should we still re-run on the release day when the tags are available? If the generated files are correct and the changes to go.mod are required, then I'll update the release docs role handbook with this required change.

reylejano commented:
@kbhawkey @tengqm I had a mistake on my last commit. I'm going to revert

reylejano commented:
I updated all the tag and versions in go.mod in commit 75ed7cc and not much change.
@kbhawkey @tengqm , I was thinking of re-running this on the release day (April 8) with the go.mod tag changes when the tags are available. What do you think?

tengqm commented:
@reylejano Please ping me when there is any issue generating the reference.

kbhawkey commented:
Hi. I asked about the .mod file because I did not expect to see the file as part of this PR.
The script builds the references in a temporary directory. Perhaps this file gets copied over after building the reference?
At this point, it looks like you will have to wait to build the other reference pages. You could push in these changes now and
recheck the build after the release is finalized?
@reylejano Did you update the reference section page (https://deploy-preview-27248--kubernetes-io-vnext-staging.netlify.app/docs/reference/)

reylejano commented:
@kbhawkey I reverted back to my initial commit and rebased my branch with `dev-1.21` which had the updated config.toml file.

https://deploy-preview-27248--kubernetes-io-vnext-staging.netlify.app/docs/reference/

kbhawkey commented:
Hi @reylejano .
Catching up here. Do you want to merge these changes and then rebuild the references (including component tool pages)
once the release is tagged?

reylejano commented:
> Hi @reylejano .
> Catching up here. Do you want to merge these changes and then rebuild the references (including component tool pages)
> once the release is tagged?

@kbhawkey, yes we can do that

sftim commented:
@reylejano it doesn't look as if CronJob has graduated to beta in this API reference. I was expecting to see it as a v1 API.
Do you know why that is?

reylejano commented:
> @reylejano it doesn't look as if CronJob has graduated to ~beta~ stable in this API reference. I was expecting to see it as a v1 API.
> Do you know why that is?

@sftim, I'm looking into this as the CronJob feature is a little more complex in that the CronJob feature is stable in 1.21 but the CronJob Controller v2 is beta in 1.21. The API is stable for 1.21 but the controller is targeted to be stable in 1.22.
